### DISCUSSION TITLE
Fine-grained access control / more authorization mechanisms

### DISCUSSION THREAD

**Discussion by morigs** [zitadel/zitadel#5090] [Ideas]
One of key Keycloak features is support for [wide variety of authorization strategies:](https://www.keycloak.org/docs/latest/authorization_services/index.html)

- RBAC
- ABAC
- UBAC
- CBAC
- etc

If I understand correctly from the docs, currently, Zitadel supports RBAC only.

Perhaps you can consider adding policy based authorization support?
Or, probably, even better solution would be to provide some way to integrate third-party mature authorization solutions (OPA, Cerbos, Ory Keto, ...)

**Comment by fforootd** [comment]
Hi There

So far we support RBAC and [Delegated Access](https://zitadel.com/docs/guides/solution-scenarios/b2b#organizations) (you can delegate Roles to other organizations).

To extend ZITADELs capabilities you can integrate with external AuthZ services like warrant.dev or cerbos.dev or basically everything that can consume ZITADELs Users and Roles.

For other concepts like ABAC we encourage users to work with our [Actions feature](https://zitadel.com/docs/concepts/features/actions).
I.e. you could create a post authentication script that checks for certain Attributes with the user and denies access if necessary.

To me it feels like the right approach to allow our users to freely select what authz engine they want to use.

Hope this helps 😁
